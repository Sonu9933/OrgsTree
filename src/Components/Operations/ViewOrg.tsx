import { useContext, useEffect } from "react";
import { OrganizationContext } from "../Context/Context";
import { Link } from "react-router-dom";
import { ActionTypes } from "../Context/ActionTypes";
import axios from "axios";
import Collapsible from "react-collapsible";
const orgGraphData: any[] = [];
const reg = new RegExp("^[0-9]+$");

const traverseJSON = (data: any) => {
  for (let k in data) {
    if (data[k] instanceof Object) {
      if (
        !reg.test(Object.keys(data).toString()) &&
        Object.keys(data).length <= 1
      ) {
        orgGraphData.push({
          name: Object.keys(data).toString(),
          position: data[k].position,
          element: "<li>",
        });
        orgGraphData.push({
          element: "<ul>",
        });
      } else {
        traverseJSON(data[k]);
      }
      traverseJSON(data[k].employees);
    } else {
      if (data instanceof Object) {
        orgGraphData.push({
          name: data.name,
          position: data.position,
          element: "<li>",
        });
        data++;
      }
    }
  }
};

const getTreeStructure = (orgData: any) => {
  let element: string = "";

  orgData.map((data: any, i: any) => {
    if (data.element === "<ul>") {
      element += "<ul>";
    } else if (data.element === "<li>") {
      element += `<li>
      <span className="node bold">${orgData[i].name} - ${orgData[i].position}</span>
    </li>`;
    }
  });
  return element;
};

export const ViewOrg = () => {
  const { state, dispatch } = useContext(OrganizationContext);
  const { organization, jsonData } = { ...state };

  useEffect(() => {
    if (organization.length === 0) {
      axios
        .get("https://jsonkeeper.com/b/OGAU")
        .then((response) => {
          traverseJSON(response.data);
          dispatch({
            type: ActionTypes.UPDATE_JSONDATA,
            payload: { data: response.data, jsonUpdate: false },
          });
          dispatch({
            type: ActionTypes.UPDATE,
            payload: { orgData: orgGraphData, jsonUpdate: false },
          });
        })
        .catch(() => {
          dispatch({
            type: ActionTypes.ERROR,
            payload: {
              error: true,
            },
          });
        });
    } else {
      traverseJSON(jsonData);
      dispatch({
        type: ActionTypes.UPDATE,
        payload: { orgData: orgGraphData, jsonUpdate: false },
      });
      dispatch({
        type: ActionTypes.ERROR,
        payload: {
          error: false,
        },
      });
    }
  }, []);

  return (
    <>
      <div className="container p-4">
        <div className="row">
          <div className="col-xl-5 col-lg-6 col-md-8 col-sm-10 mx-auto text-center">
            <h1 className="text-center display-4 py-2 text-truncate text-primary text-underline">
              <b>Organization Chart</b>
            </h1>
          </div>
        </div>
        <div className="row">
          <div className="col-lg-8 col-md-8 col-sm-10 mx-auto">
            <p className="mx-auto text-center Info link ">
              <b>
                {" "}
                Below data is based on the date received from end point
                (https://jsonkeeper.com/b/OGAU). To update the it, please upload
                new json by clicking on Upload JSON link{" "}
              </b>
            </p>
            <div
              className="p-4 rounded"
              style={{ minHeight: 200, border: "1px solid black" }}
            >
              <Collapsible trigger="Click to view">
                <ul
                  id="org-tree"
                  dangerouslySetInnerHTML={{
                    __html: getTreeStructure(organization),
                  }}
                />
              </Collapsible>
            </div>
          </div>
        </div>
      </div>
      <div className="row">
        <div className="col-xl-5 col-lg-6 col-md-8 col-sm-10 mx-auto text-center form p-4">
          Click here to Update hierarchy :
          <b>
            <Link to="/UpdateOrg" className="px-2">
              Upload JSON
            </Link>
          </b>
        </div>
      </div>
    </>
  );
};
