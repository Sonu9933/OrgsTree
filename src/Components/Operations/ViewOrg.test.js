import React from "react";
import { ViewOrg } from "./ViewOrg";
import { renderWithContext } from "../Test/TestHelper";
describe("View Org", () => {
  test("Should render ViewOrg", () => {
    const component = renderWithContext(<ViewOrg />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
