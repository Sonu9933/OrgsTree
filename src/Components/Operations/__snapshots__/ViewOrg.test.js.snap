// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Org Should render ViewOrg 1`] = `
Array [
  <div
    className="container p-4"
  >
    <div
      className="row"
    >
      <div
        className="col-xl-5 col-lg-6 col-md-8 col-sm-10 mx-auto text-center"
      >
        <h1
          className="text-center display-4 py-2 text-truncate text-primary text-underline"
        >
          <b>
            Organization Chart
          </b>
        </h1>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-lg-8 col-md-8 col-sm-10 mx-auto"
      >
        <p
          className="mx-auto text-center Info link "
        >
          <b>
             
            Below data is based on the date received from end point (https://jsonkeeper.com/b/OGAU). To update the it, please upload new json by clicking on Upload JSON link
             
          </b>
        </p>
        <div
          className="p-4 rounded"
          style={
            Object {
              "border": "1px solid black",
              "minHeight": 200,
            }
          }
        >
          <div
            className="Collapsible"
          >
            <span
              aria-controls="collapsible-content-1631362601322"
              aria-disabled={false}
              aria-expanded={false}
              className="Collapsible__trigger is-closed"
              id="collapsible-trigger-1631362601322"
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              style={null}
              tabIndex={null}
            >
              Click to view
            </span>
            <div
              aria-labelledby="collapsible-trigger-1631362601322"
              className="Collapsible__contentOuter"
              hidden={false}
              id="collapsible-content-1631362601322"
              onTransitionEnd={[Function]}
              role="region"
              style={
                Object {
                  "WebkitTransition": "height 400ms linear",
                  "height": 0,
                  "msTransition": "height 400ms linear",
                  "overflow": "hidden",
                  "transition": "height 400ms linear",
                }
              }
            >
              <div
                className="Collapsible__contentInner"
              >
                <ul
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "",
                    }
                  }
                  id="org-tree"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-xl-5 col-lg-6 col-md-8 col-sm-10 mx-auto text-center form p-4"
    >
      Click here to Update hierarchy :
      <b>
        <a
          className="px-2"
          href="/UpdateOrg"
          onClick={[Function]}
        >
          Upload JSON
        </a>
      </b>
    </div>
  </div>,
]
`;
